src/sql/azdata.proposed.d.ts

```diff
@@ -432,6 +432,53 @@ declare module 'azdata' {
 		azurePortalEndpoint?: string;
 	}
 
+
+	export namespace diagnostics {
+		/**
+		 * Represents a diagnostics provider of accounts.
+		 */
+		export interface ErrorDiagnosticsProviderMetadata {
+			/**
+			 * The name of the provider (ex. a connection provider) that the diagnostics provider will handle errors for.
+			 */
+			id: string;
+		}
+
+		export interface ConnectionDiagnosticsResult {
+			/**
+			 * Status indicating if the error was handled or not.
+			 */
+			handled: boolean,
+			/**
+			 * The altered connection options after error handling.
+			 */
+			options: { [name: string]: any };
+		}
+
+		/**
+		 * Diagnostics object for handling errors for a provider.
+		 */
+		export interface ErrorDiagnosticsProvider {
+			/**
+			 * Called when a connection error occurs, allowing the provider to optionally handle the error and fix any issues before continuing with completing the connection.
+			 * @param errorCode The error code of the connection error.
+			 * @param errorMessage The error message of the connection error.
+			 * @param connection The connection profile that caused the error.
+			 * @returns ConnectionDiagnosticsResult object containing error handling success status (boolean), and the changed connection options of the profile.
```

Comment:
Better to make this a bit more generic so if you add other stuff to the result later on you don't have to remember to update this

```suggestion
			 * @returns ConnectionDiagnosticsResult The result from the provider for whether the error was handled.
```